File: corpus/bench_dot.cpp
// C++ twin of bench_dot.lu — idiomatic style, heap vectors.
#include <cstdio>
#include <vector>

double dot(const std::vector<double>& a, const std::vector<double>& b, long n) {
  double s = 0.0;
  for (long i = 0; i < n; i++) s += a[i] * b[i];
  return s;
}

int main() {
  const long n = 2000000;
  std::vector<double> a(n), b(n);
  for (long i = 0; i < n; i++) {
    a[i] = i * 0.000001;
    b[i] = i * 0.000002;
  }
  double acc = 0.0;
  for (int r = 0; r < 200; r++) acc += dot(a, b, n);
  printf("acc: %.17g\n", acc);
}

File: corpus/bench_dot.rs
// Rust twin of bench_dot.lu — idiomatic owned vectors and a scalar iterator.

fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b).map(|(x, y)| x * y).sum()
}

fn main() {
    const N: usize = 2_000_000;
    let a = (0..N).map(|i| i as f64 * 0.000001).collect::<Vec<_>>();
    let b = (0..N).map(|i| i as f64 * 0.000002).collect::<Vec<_>>();
    let mut acc = 0.0;
    for _ in 0..20 {
        acc += dot(&a, &b);
    }
    println!("acc: {acc:.17}");
}
